Cannot infer generic receiver types

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
#[lang = "sized"]
pub trait Sized {}

struct S<T>(T);

impl S<u8> {
    fn foo<U>() {}
}

fn main() {
    S::foo::<i32>();
}

```

I expected to see this happen:
generic types of S is inferred and monomorphized to u32

Instead, this happened: 
```
$ ./crab1  b.rs
b.rs:11:14: error: generic item takes at most 1 type arguments but 1 were supplied
    7 |     fn foo<U>() {}
      |            ~  
......
   11 |     S::foo::<i32>();
      |              ^~~
b.rs:11:5: error: Failed to resolve expression of function call
   11 |     S::foo::<i32>();
      |     ^

Analyzing compilation unit

Time variable                                   usr           sys          wall           GGC
 TOTAL                              :   0.00          0.00          0.00          147k
Extra diagnostic checks enabled; compiler may run slowly.
Configure with --enable-checking=release to disable checks.
```

### Meta

- master

--- a/gcc/testsuite/rust/compile/issue-2782.rs
+++ b/gcc/testsuite/rust/compile/issue-2782.rs
@@ -1,0 +1,12 @@
+#[lang = "sized"]
+pub trait Sized {}
+
+struct S<T>(T);
+
+impl S<u8> {
+    fn foo<U>() {}
+}
+
+fn main() {
+    S::foo::<i32>();
+}

